### PR TITLE
ENH: Add ITK C++ and Python CI Reusable Workflows

### PR DESCRIPTION
**Pull request by tbirdso**
Migrate ITK C++ and Python Github CI procedures into reusable workflows that can be called and run by ITK external modules.

Partially addresses [https://github.com/InsightSoftwareConsortium/ITKModuleTemplate/issues/131](https://github.com/InsightSoftwareConsortium/ITKModuleTemplate/issues/131).

Caveats:
1. Reusable workflows are only designed and tested for simple remote modules with no extra build steps.
2. Passing custom CMake arguments to a remote module build is feasible but untested. Custom CMake arguments should be considered not supported for this first iteration.
3. Remote modules depending on other remote modules or libraries are not currently supported. Modules with extra dependencies are responsible for implementing their own CI pipelines. The workflows here have been forked from `ITKModuleTemplate` and can in turn be forked as a starting point for a custom pipeline in a remote module repository.

### CODE REVIEW

File: .github/workflows/build-test-cxx.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      shell: bash
+      run: |
+        operating_system="${{ matrix.os }}"
+        cat > dashboard.cmake << EOF
+        set(CTEST_SITE "GitHubActions")
+        file(TO_CMAKE_PATH "\$ENV{GITHUB_WORKSPACE}/.." CTEST_DASHBOARD_ROOT)
+        file(TO_CMAKE_PATH "\$ENV{GITHUB_WORKSPACE}/" CTEST_SOURCE_DIRECTORY)
+        file(TO_CMAKE_PATH "\$ENV{GITHUB_WORKSPACE}/../build" CTEST_BINARY_DIRECTORY)
+        set(dashboard_source_name "${GITHUB_REPOSITORY}")
+        if(ENV{GITHUB_REF} MATCHES "master")
```
[the remark this answers is not in the document]
Comment (tbirdso) [reply .github/workflows/build-test-cxx.yml:95]: Yes, good catch, will add

[the remark this answers is not in the document]
Comment (N-Dekker) [reply .github/workflows/build-test-cxx.yml:95]: Just wondering, should it be `GITHUB_REF` or `GITHUB_REF_NAME`? `GITHUB_REF_NAME` appears to be the shorter name, while `GITHUB_REF` would be of the form "refs/heads/feature-branch-1" : https://docs.github.com/en/actions/learn-github-actions/environment-variables

[the remark this answers is not in the document]
Comment (tbirdso) [reply .github/workflows/build-test-cxx.yml]: @N-Dekker Thanks for your patience, I agree that `GITHUB_REF_NAME` is the correct value to check. It looks like we've been using `GITHUB_REF` in other pipelines, i.e. all module build results are currently filed under `Experimental`. I've fixed the conditional here, will continue to use the expanded name in the build name associated on CDash.

**Comment by tbirdso** [comment]
Observed passing CI on ITKSplitComponents: https://github.com/InsightSoftwareConsortium/ITKSplitComponents/actions/runs/3207298168/jobs/5242405654

Changes have been squashed. Ready for review.

File: .github/workflows/build-test-cxx.yml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            cmake-build-type: "Release"
+          - os: macos-11
+            c-compiler: "clang"
+            cxx-compiler: "clang++"
+            cmake-build-type: "MinSizeRel"
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Python 3.8
```
Comment (dzenanz) [inline .github/workflows/build-test-cxx.yml:41]: Should we have 3.9 here? [ref](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.3rc04/.circleci/config.yml#L18)

File: .github/workflows/build-test-cxx.yml [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.8
```
Comment (dzenanz) [inline .github/workflows/build-test-cxx.yml:44]: Also here https://github.com/InsightSoftwareConsortium/ITK/blob/v5.3rc04/Testing/ContinuousIntegration/AzurePipelinesLinuxPython.yml#L29

**Review by dzenanz — APPROVED** [review]
Code looks good. The big test is whether it works 😄

**Comment by tbirdso** [comment]
Thanks @dzenanz . Note that development tests in ITKSplitComponents using these workflows are passing: https://github.com/InsightSoftwareConsortium/ITKSplitComponents/pull/63

File: .github/workflows/build-test-cxx.yml (hunk 2 above)
Comment (tbirdso) [reply .github/workflows/build-test-cxx.yml:41]: I have submitted https://github.com/InsightSoftwareConsortium/ITKRemoteModuleBuildTestPackageAction/issues/4 to track.

File: .github/workflows/build-test-cxx.yml (hunk 3 above)
Comment (tbirdso) [reply .github/workflows/build-test-cxx.yml:44]: Submitted https://github.com/InsightSoftwareConsortium/ITKRemoteModuleBuildTestPackageAction/issues/4 to track

**Comment by tbirdso** [comment]
Going ahead and merging so that reusable workflows can be tagged and referenced in remote module CI. Additional feedback can be added as discussions here or in the issue tracker.

**Pull request merged**